Implement unit tests

Running `npm run lint` produces the following error:

``` sh
/projects/redux-form-validation/src/basic-validations.js
  1:2  error  Parsing error: Illegal export declaration

/projects/redux-form-validation/src/form-messages/index.js
  1:2  error  Parsing error: Illegal import declaration

/projects/redux-form-validation/src/index.js
  1:2  error  Parsing error: Illegal import declaration

/projects/redux-form-validation/src/validation.js
  1:2  error  Parsing error: Illegal import declaration

✖ 4 problems (4 errors, 0 warnings)
```

This also means that running `npm run test` fails.

Once this is fixed, some unit tests can be written:

* [x] required
* [x] minLength
* [x] maxLength
* [x] email
* [x] min
* [x] max
* [x] pattern
* [x] equalTo
* [x] oneOf
* [x] url
* [x] promise
* [x] digits
* [x] creditcard

--- a/.eslintrc
+++ b/.eslintrc
@@ -1,0 +1,3 @@
+{
+  "parser": "babel-eslint"
+}

--- a/.gitignore
+++ b/.gitignore
@@ -1,4 +1,5 @@
 .idea
+*.iml
 node_modules
 dist
 lib

--- a/package.json
+++ b/package.json
@@ -34,7 +34,7 @@
     "clean": "rimraf dist lib",
     "lint": "eslint src",
     "prepublish": "npm run clean && npm run build",
-    "test": "npm run lint && mocha --compilers js:babel/register --recursive 'src/**/__tests__/*' --require src/__tests__/setup.js"
+    "test": "npm run lint && mocha --compilers js:babel/register --recursive 'src/**/__tests__/*.spec.js' --require src/__tests__/setup.js"
   },
   "devDependencies": {
     "babel": "^5.8.23",

--- a/package.json
+++ b/package.json
@@ -66,6 +66,7 @@
     }
   ],
   "dependencies": {
-    "is-promise": "^2.1.0"
+    "is-promise": "^2.1.0",
+    "valid-url": "^1.0.9"
   }
 }

--- a/src/__tests__/creditcard.spec.js
+++ b/src/__tests__/creditcard.spec.js
@@ -27,7 +27,7 @@ const generateInvalid = (pieces, separator) => {
 };
 
 // Note: each validator returns `true` when the `value` is invalid
-describe.only('Validator: creditcard', function () {
+describe('Validator: creditcard', function () {
   describe('Should be valid', () => {
     describe('With no separator', () => {
       it('When `prop` is valid card number', () => {

--- a/src/__tests__/creditcard.spec.js
+++ b/src/__tests__/creditcard.spec.js
@@ -1,0 +1,96 @@
+import assert from 'assert';
+import { creditcard } from '../basic-validations';
+// Sample valid CC numbers, split up into 4 separate pieces
+// Source: http://www.getcreditcardnumbers.com/
+const cards = [
+  { type: 'Amex', pieces: [3483, 2314, 1931, 128] },
+  { type: 'Diners', pieces: [3017, 5465, 5838, 90] },
+  { type: 'Discover', pieces: [6011, 6855, 1062, 3624] },
+  { type: 'JCB', pieces: [3528, 7139, 4254, 4326] },
+  { type: 'JCB 15', pieces: [2100, 1991, 2694, 367] },
+  { type: 'MasterCard', pieces: [5519, 4093, 1282, 6387] },
+  { type: 'Visa 13', pieces: [4225, 4269, 4946, 6] },
+  { type: 'Visa 16', pieces: [4532, 6628, 9297, 4839] },
+];
+
+const generate = (pieces, separator = '') =>
+  pieces.join(separator);
+
+const generateValid = (pieces, separator) =>
+  generate(pieces, separator);
+
+const generateInvalid = (pieces, separator) => {
+  // only one digit in the card number needs to be altered in order for it to be invalid
+  const invalidCardNumber = [pieces[0], pieces[1], pieces[2], pieces[3] + 1];
+
+  return generate(invalidCardNumber, separator)
+};
+
+// Note: each validator returns `true` when the `value` is invalid
+describe.only('Validator: creditcard', function () {
+  describe('Should be valid', () => {
+    describe('With no separator', () => {
+      it('When `prop` is valid card number', () => {
+        cards.forEach((card) => {
+          console.log('Card type:', card.type);
+          assert(creditcard(true, generateValid(card.pieces)) === false);
+        });
+      });
+    });
+    describe('With space separator', () => {
+      it('When `prop` is valid card number', () => {
+        cards.forEach((card) => {
+          console.log('Card type:', card.type);
+          assert(creditcard(true, generateValid(card.pieces, ' ')) === false);
+        });
+      });
+    });
+    describe('With "-" separator', () => {
+      it('When `prop` is valid card number', () => {
+        cards.forEach((card) => {
+          console.log('Card type:', card.type);
+          assert(creditcard(true, generateValid(card.pieces, '-')) === false);
+        });
+      });
+    });
+  });
+  describe('Should be invalid', () => {
+    describe('With no separator', () => {
+      it('When `prop` is valid card number', () => {
+        cards.forEach((card) => {
+          console.log('Card type:', card.type);
+          assert(creditcard(true, generateInvalid(card.pieces)) === true);
+        });
+      });
+    });
+    describe('With space separator', () => {
+      it('When `prop` is valid card number', () => {
+        cards.forEach((card) => {
+          console.log('Card type:', card.type);
+          assert(creditcard(true, generateInvalid(card.pieces, ' ')) === true);
+        });
+      });
+    });
+    describe('With "-" separator', () => {
+      it('When `prop` is valid card number', () => {
+        cards.forEach((card) => {
+          console.log('Card type:', card.type);
+          assert(creditcard(true, generateInvalid(card.pieces, '-')) === true);
+        });
+      });
+    });
+    describe('With non-number values', () => {
+      it('When `prop` contains letters', () => {
+        assert(creditcard(true, 'blah99') === true);
+      });
+    });
+    describe('With number too long', () => {
+      it('When `prop` contains number with length 1 longer than valid number', () => {
+        cards.forEach((card) => {
+          console.log('Card type:', card.type);
+          assert(creditcard(true, `${generateValid(card.pieces)}9`) === true);
+        });
+      });
+    });
+  });
+});

--- a/src/__tests__/digits.spec.js
+++ b/src/__tests__/digits.spec.js
@@ -1,0 +1,31 @@
+import assert from 'assert';
+import { digits } from '../basic-validations';
+
+// Note: each validator returns `true` when the `value` is invalid
+describe('Validator: digits', function () {
+  describe('Should be valid', () => {
+    it('When `prop` is "999"', () => {
+      assert(digits(true, '999') === false);
+    });
+    it('When `prop` is 999', () => {
+      assert(digits(true, 999) === false);
+    });
+  });
+  describe('Should be invalid', () => {
+    it('When `prop` is "blah"', () => {
+      assert(digits({}, 'blah') === true);
+    });
+    it('When `prop` is "blah77"', () => {
+      assert(digits({}, 'blah77') === true);
+    });
+    it('When `prop` is "4,000"', () => {
+      assert(digits({}, '4,000') === true);
+    });
+    it('When `prop` is "77.7"', () => {
+      assert(digits({}, '77.7') === true);
+    });
+    it('When `prop` is 77.7', () => {
+      assert(digits({}, 77.7) === true);
+    });
+  });
+});

--- a/src/__tests__/email.spec.js
+++ b/src/__tests__/email.spec.js
@@ -3,33 +3,33 @@ import { email } from '../basic-validations';
 
 // Note: each validator returns `true` when the `value` is invalid
 describe('Validator: email', function validateMaxLength() {
-  it('should be invalid when `prop` is `x`', function () {
-    assert(email({}, `x`, true) === true);
+  it('should be invalid when `prop` is "x"', function () {
+    assert(email({}, 'x', true) === true);
   });
-  it('should be invalid when `prop` is `x@`', function () {
-    assert(email({}, `x@`, true) === true);
+  it('should be invalid when `prop` is "x@"', function () {
+    assert(email({}, 'x@', true) === true);
   });
-  it('should be invalid when `prop` is `@x`', function () {
-    assert(email({}, `@x`, true) === true);
+  it('should be invalid when `prop` is "@x"', function () {
+    assert(email({}, '@x', true) === true);
   });
-  it('should be invalid when `prop` is `x@x`', function () {
+  it('should be invalid when `prop` is "x@x"', function () {
     // Note: we are assuming that the domain name will contain at least one '.'.
     // So `localhost`, for example, will not be considered a valid domain name.
-    assert(email({}, `x@x`, true) === true);
+    assert(email({}, 'x@x', true) === true);
   });
-  it('should be invalid when `prop` is `x@x.`', function () {
-    assert(email({}, `x@x.`, true) === true);
+  it('should be invalid when `prop` is "x@x."', function () {
+    assert(email({}, 'x@x.', true) === true);
   });
-  it('should be valid when `prop` is `x@x.x`', function () {
-    assert(email({}, `x@x.x`, true) === false);
+  it('should be valid when `prop` is "x@x.x"', function () {
+    assert(email({}, 'x@x.x', true) === false);
   });
-  it('should be valid when `prop` is `x@x.xx`', function () {
-    assert(email({}, `x@x.xx`, true) === false);
+  it('should be valid when `prop` is "x@x.xx"', function () {
+    assert(email({}, 'x@x.xx', true) === false);
   });
-  it('should be valid when `prop` is `x@xx.xx`', function () {
-    assert(email({}, `x@xx.xx`, true) === false);
+  it('should be valid when `prop` is "x@xx.xx"', function () {
+    assert(email({}, 'x@xx.xx', true) === false);
   });
-  it('should be valid when `prop` is `x@192.16.0.10`', function () {
-    assert(email({}, `x@192.16.0.10`, true) === false);
+  it('should be valid when `prop` is "x@192.16.0.10"', function () {
+    assert(email({}, 'x@192.16.0.10', true) === false);
   });
 });

--- a/src/__tests__/email.spec.js
+++ b/src/__tests__/email.spec.js
@@ -1,0 +1,35 @@
+import assert from 'assert';
+import { email } from '../basic-validations';
+
+// Note: each validator returns `true` when the `value` is invalid
+describe('Validator: email', function validateMaxLength() {
+  it('should be invalid when `prop` is `x`', function () {
+    assert(email({}, `x`, true) === true);
+  });
+  it('should be invalid when `prop` is `x@`', function () {
+    assert(email({}, `x@`, true) === true);
+  });
+  it('should be invalid when `prop` is `@x`', function () {
+    assert(email({}, `@x`, true) === true);
+  });
+  it('should be invalid when `prop` is `x@x`', function () {
+    // Note: we are assuming that the domain name will contain at least one '.'.
+    // So `localhost`, for example, will not be considered a valid domain name.
+    assert(email({}, `x@x`, true) === true);
+  });
+  it('should be invalid when `prop` is `x@x.`', function () {
+    assert(email({}, `x@x.`, true) === true);
+  });
+  it('should be valid when `prop` is `x@x.x`', function () {
+    assert(email({}, `x@x.x`, true) === false);
+  });
+  it('should be valid when `prop` is `x@x.xx`', function () {
+    assert(email({}, `x@x.xx`, true) === false);
+  });
+  it('should be valid when `prop` is `x@xx.xx`', function () {
+    assert(email({}, `x@xx.xx`, true) === false);
+  });
+  it('should be valid when `prop` is `x@192.16.0.10`', function () {
+    assert(email({}, `x@192.16.0.10`, true) === false);
+  });
+});

--- a/src/__tests__/email.spec.js
+++ b/src/__tests__/email.spec.js
@@ -2,7 +2,7 @@ import assert from 'assert';
 import { email } from '../basic-validations';
 
 // Note: each validator returns `true` when the `value` is invalid
-describe('Validator: email', function validateMaxLength() {
+describe('Validator: email', function () {
   it('should be invalid when `prop` is "x"', function () {
     assert(email({}, 'x', true) === true);
   });

--- a/src/__tests__/equalTo.spec.js
+++ b/src/__tests__/equalTo.spec.js
@@ -2,7 +2,7 @@ import assert from 'assert';
 import { equalTo } from '../basic-validations';
 
 // Note: each validator returns `true` when the `value` is invalid
-describe('Validator: equalTo', function validateMaxLength() {
+describe('Validator: equalTo', function () {
   it('should be valid when `value` (string) equals `prop` (string)', function () {
     assert(equalTo({}, 'matcher', 'matcher') === false);
   });

--- a/src/__tests__/equalTo.spec.js
+++ b/src/__tests__/equalTo.spec.js
@@ -1,0 +1,18 @@
+import assert from 'assert';
+import { equalTo } from '../basic-validations';
+
+// Note: each validator returns `true` when the `value` is invalid
+describe('Validator: equalTo', function validateMaxLength() {
+  it('should be valid when `value` (string) equals `prop` (string)', function () {
+    assert(equalTo({}, 'matcher', 'matcher') === false);
+  });
+  it('should be invalid when `value` (string) does not equal `prop` (string)', function () {
+    assert(equalTo({}, 'non-matcher', 'matcher') === true);
+  });
+  it('should be valid when `value` (number) equals `prop` (number)', function () {
+    assert(equalTo({}, 77, 77) === false);
+  });
+  it('should be valid when `value` (number) does not equal `prop` (number)', function () {
+    assert(equalTo({}, 77, 88) === true);
+  });
+});

--- a/src/__tests__/max.spec.js
+++ b/src/__tests__/max.spec.js
@@ -2,7 +2,7 @@ import assert from 'assert';
 import { max } from '../basic-validations';
 
 // Note: each validator returns `true` when the `value` is invalid
-describe('Validator: max', function validateMinLength() {
+describe('Validator: max', function () {
   it('should be valid when `value` is undefined and `prop` is 3', function () {
     assert(max({}, undefined, 3) === false);
   });

--- a/src/__tests__/max.spec.js
+++ b/src/__tests__/max.spec.js
@@ -1,0 +1,27 @@
+import assert from 'assert';
+import { max } from '../basic-validations';
+
+// Note: each validator returns `true` when the `value` is invalid
+describe('Validator: max', function validateMinLength() {
+  it('should be valid when `value` is undefined and `prop` is 3', function () {
+    assert(max({}, undefined, 3) === false);
+  });
+  it('should be valid when `value` is "99" and `prop` is 100', function () {
+    assert(max({}, '99', 100) === false);
+  });
+  it('should be valid when `value` is "100" and `prop` is 100', function () {
+    assert(max({}, '100', 100) === false);
+  });
+  it('should be invalid when `value` is "101" and `prop` is 100', function () {
+    assert(max({}, '101', 100) === true);
+  });
+  it('should be valid when `value` is 99 and `prop` is 100', function () {
+    assert(max({}, 99, 100) === false);
+  });
+  it('should be valid when `value` is 100 and `prop` is 100', function () {
+    assert(max({}, 100, 100) === false);
+  });
+  it('should be invalid when `value` is 101 and `prop` is 100', function () {
+    assert(max({}, 101, 100) === true);
+  });
+});

--- a/src/__tests__/maxLength.spec.js
+++ b/src/__tests__/maxLength.spec.js
@@ -3,16 +3,16 @@ import { maxLength } from '../basic-validations';
 
 // Note: each validator returns `true` when the `value` is invalid
 describe('Validator: maxLength', function validateMaxLength() {
-  it('should be valid when `value` is `undefined` and `prop` is 3', function () {
+  it('should be valid when `value` is undefined and `prop` is 3', function () {
     assert(maxLength({}, undefined, 3) === false);
   });
-  it('should be invalid when `value` is `1234` and `prop` is 3', function () {
+  it('should be invalid when `value` is "1234" and `prop` is 3', function () {
     assert(maxLength({}, '1234', 3) === true);
   });
-  it('should be valid when `value` is `1234` and `prop` is 4', function () {
+  it('should be valid when `value` is "1234" and `prop` is 4', function () {
     assert(maxLength({}, '1234', 4) === false);
   });
-  it('should be valid when `value` is `1234` and `prop` is 5', function () {
+  it('should be valid when `value` is "1234" and `prop` is 5', function () {
     assert(maxLength({}, '1234', 5) === false);
   });
 });

--- a/src/__tests__/maxLength.spec.js
+++ b/src/__tests__/maxLength.spec.js
@@ -2,7 +2,7 @@ import assert from 'assert';
 import { maxLength } from '../basic-validations';
 
 // Note: each validator returns `true` when the `value` is invalid
-describe('Validator: maxLength', function validateMaxLength() {
+describe('Validator: maxLength', function () {
   it('should be valid when `value` is undefined and `prop` is 3', function () {
     assert(maxLength({}, undefined, 3) === false);
   });

--- a/src/__tests__/maxLength.spec.js
+++ b/src/__tests__/maxLength.spec.js
@@ -1,0 +1,18 @@
+import assert from 'assert';
+import { maxLength } from '../basic-validations';
+
+// Note: each validator returns `true` when the `value` is invalid
+describe('Validator: maxLength', function validateMaxLength() {
+  it('should be valid when `value` is `undefined` and `prop` is 3', function () {
+    assert(maxLength({}, undefined, 3) === false);
+  });
+  it('should be invalid when `value` is `1234` and `prop` is 3', function () {
+    assert(maxLength({}, '1234', 3) === true);
+  });
+  it('should be valid when `value` is `1234` and `prop` is 4', function () {
+    assert(maxLength({}, '1234', 4) === false);
+  });
+  it('should be valid when `value` is `1234` and `prop` is 5', function () {
+    assert(maxLength({}, '1234', 5) === false);
+  });
+});

--- a/src/__tests__/min.spec.js
+++ b/src/__tests__/min.spec.js
@@ -2,7 +2,7 @@ import assert from 'assert';
 import { min } from '../basic-validations';
 
 // Note: each validator returns `true` when the `value` is invalid
-describe('Validator: min', function validateMinLength() {
+describe('Validator: min', function () {
   it('should be valid when `value` is undefined and `prop` is 3', function () {
     assert(min({}, undefined, 3) === false);
   });

--- a/src/__tests__/min.spec.js
+++ b/src/__tests__/min.spec.js
@@ -1,0 +1,27 @@
+import assert from 'assert';
+import { min } from '../basic-validations';
+
+// Note: each validator returns `true` when the `value` is invalid
+describe('Validator: min', function validateMinLength() {
+  it('should be valid when `value` is undefined and `prop` is 3', function () {
+    assert(min({}, undefined, 3) === false);
+  });
+  it('should be invalid when `value` is "99" and `prop` is 100', function () {
+    assert(min({}, '99', 100) === true);
+  });
+  it('should be valid when `value` is "100" and `prop` is 100', function () {
+    assert(min({}, '100', 100) === false);
+  });
+  it('should be valid when `value` is "101" and `prop` is 100', function () {
+    assert(min({}, '101', 100) === false);
+  });
+  it('should be invalid when `value` is 99 and `prop` is 100', function () {
+    assert(min({}, 99, 100) === true);
+  });
+  it('should be valid when `value` is 100 and `prop` is 100', function () {
+    assert(min({}, 100, 100) === false);
+  });
+  it('should be valid when `value` is 101 and `prop` is 100', function () {
+    assert(min({}, 101, 100) === false);
+  });
+});

--- a/src/__tests__/minLength.spec.js
+++ b/src/__tests__/minLength.spec.js
@@ -2,7 +2,7 @@ import assert from 'assert';
 import { minLength } from '../basic-validations';
 
 // Note: each validator returns `true` when the `value` is invalid
-describe('Validator: minLength', function validateMinLength() {
+describe('Validator: minLength', function () {
   it('should be valid when `value` is undefined and `prop` is 3', function () {
     assert(minLength({}, undefined, 3) === false);
   });

--- a/src/__tests__/minLength.spec.js
+++ b/src/__tests__/minLength.spec.js
@@ -3,16 +3,16 @@ import { minLength } from '../basic-validations';
 
 // Note: each validator returns `true` when the `value` is invalid
 describe('Validator: minLength', function validateMinLength() {
-  it('should be false when `value` is `undefined` and `prop` is 3', function () {
+  it('should be valid when `value` is `undefined` and `prop` is 3', function () {
     assert(minLength({}, undefined, 3) === false);
   });
-  it('should be false when `value` is `1234` and `prop` is 3', function () {
+  it('should be valid when `value` is `1234` and `prop` is 3', function () {
     assert(minLength({}, '1234', 3) === false);
   });
-  it('should be false when `value` is `1234` and `prop` is 4', function () {
+  it('should be valid when `value` is `1234` and `prop` is 4', function () {
     assert(minLength({}, '1234', 4) === false);
   });
-  it('should be false when `value` is `1234` and `prop` is 5', function () {
+  it('should be invalid when `value` is `1234` and `prop` is 5', function () {
     assert(minLength({}, '1234', 5) === true);
   });
 });

--- a/src/__tests__/minLength.spec.js
+++ b/src/__tests__/minLength.spec.js
@@ -1,0 +1,18 @@
+import assert from 'assert';
+import { minLength } from '../basic-validations';
+
+// Note: each validator returns `true` when the `value` is invalid
+describe('Validator: minLength', function validateMinLength() {
+  it('should be false when `value` is `undefined` and `prop` is 3', function () {
+    assert(minLength({}, undefined, 3) === false);
+  });
+  it('should be false when `value` is `1234` and `prop` is 3', function () {
+    assert(minLength({}, '1234', 3) === false);
+  });
+  it('should be false when `value` is `1234` and `prop` is 4', function () {
+    assert(minLength({}, '1234', 4) === false);
+  });
+  it('should be false when `value` is `1234` and `prop` is 5', function () {
+    assert(minLength({}, '1234', 5) === true);
+  });
+});

--- a/src/__tests__/minLength.spec.js
+++ b/src/__tests__/minLength.spec.js
@@ -3,16 +3,16 @@ import { minLength } from '../basic-validations';
 
 // Note: each validator returns `true` when the `value` is invalid
 describe('Validator: minLength', function validateMinLength() {
-  it('should be valid when `value` is `undefined` and `prop` is 3', function () {
+  it('should be valid when `value` is undefined and `prop` is 3', function () {
     assert(minLength({}, undefined, 3) === false);
   });
-  it('should be valid when `value` is `1234` and `prop` is 3', function () {
+  it('should be valid when `value` is "1234" and `prop` is 3', function () {
     assert(minLength({}, '1234', 3) === false);
   });
-  it('should be valid when `value` is `1234` and `prop` is 4', function () {
+  it('should be valid when `value` is "1234" and `prop` is 4', function () {
     assert(minLength({}, '1234', 4) === false);
   });
-  it('should be invalid when `value` is `1234` and `prop` is 5', function () {
+  it('should be invalid when `value` is "1234" and `prop` is 5', function () {
     assert(minLength({}, '1234', 5) === true);
   });
 });

--- a/src/__tests__/oneOf.spec.js
+++ b/src/__tests__/oneOf.spec.js
@@ -1,0 +1,36 @@
+import assert from 'assert';
+import { oneOf } from '../basic-validations';
+
+// Note: each validator returns `true` when the `value` is invalid
+describe('Validator: oneOf', function validateMaxLength() {
+  it('should be valid when `value` (string) is listed in `prop`', function () {
+    assert(oneOf({}, 'bb', ['aa', 'bb', 'cc']) === false);
+  });
+  it('should be invalid when `value` (string) is not listed in `prop`', function () {
+    assert(oneOf({}, 'zz', ['aa', 'bb', 'cc']) === true);
+  });
+  it('should be valid when `value` (number) is listed in `prop`', function () {
+    assert(oneOf({}, 13, [11, 12, 13]) === false);
+  });
+  it('should be invalid when `value` (number) is not in `prop`', function () {
+    assert(oneOf({}, 99, [11, 12, 13]) === true);
+  });
+  it('should throw TypeError when `prop` is boolean', function () {
+    assert.throws(
+      () => oneOf({}, 'z', true),
+      TypeError
+    );
+  });
+  it('should throw TypeError when `prop` is object', function () {
+    assert.throws(
+      () => oneOf({}, 'z', {}),
+      TypeError
+    );
+  });
+  it('should throw TypeError when `prop` is number', function () {
+    assert.throws(
+      () => oneOf({}, 'z', 99),
+      TypeError
+    );
+  });
+});

--- a/src/__tests__/oneOf.spec.js
+++ b/src/__tests__/oneOf.spec.js
@@ -2,7 +2,7 @@ import assert from 'assert';
 import { oneOf } from '../basic-validations';
 
 // Note: each validator returns `true` when the `value` is invalid
-describe('Validator: oneOf', function validateMaxLength() {
+describe('Validator: oneOf', function () {
   it('should be valid when `value` (string) is listed in `prop`', function () {
     assert(oneOf({}, 'bb', ['aa', 'bb', 'cc']) === false);
   });

--- a/src/__tests__/pattern.spec.js
+++ b/src/__tests__/pattern.spec.js
@@ -2,7 +2,7 @@ import assert from 'assert';
 import { pattern } from '../basic-validations';
 
 // Note: each validator returns `true` when the `value` is invalid
-describe('Validator: pattern', function validateMaxLength() {
+describe('Validator: pattern', function () {
   it('should be valid when `value` matches `prop` regex', function () {
     assert(pattern({}, '__matcher__', /.*matcher.*/) === false);
   });

--- a/src/__tests__/pattern.spec.js
+++ b/src/__tests__/pattern.spec.js
@@ -9,25 +9,25 @@ describe('Validator: pattern', function validateMaxLength() {
   it('should be invalid when `value` does not match `prop` regex', function () {
     assert(pattern({}, '__doesNotMatch__', /.*matcher.*/) === true);
   });
-  it('should be invalid when `prop` is boolean', function () {
+  it('should throw TypeError when `prop` is boolean', function () {
     assert.throws(
       () => pattern({}, 'z', true),
       TypeError
     );
   });
-  it('should be invalid when `prop` is string', function () {
+  it('should throw TypeError when `prop` is string', function () {
     assert.throws(
       () => pattern({}, 'z', 'a nice string'),
       TypeError
     );
   });
-  it('should be invalid when `prop` is object', function () {
+  it('should throw TypeError when `prop` is object', function () {
     assert.throws(
       () => pattern({}, 'z', {}),
       TypeError
     );
   });
-  it('should be invalid when `prop` is number', function () {
+  it('should throw TypeError when `prop` is number', function () {
     assert.throws(
       () => pattern({}, 'z', 99),
       TypeError

--- a/src/__tests__/pattern.spec.js
+++ b/src/__tests__/pattern.spec.js
@@ -1,0 +1,36 @@
+import assert from 'assert';
+import { pattern } from '../basic-validations';
+
+// Note: each validator returns `true` when the `value` is invalid
+describe('Validator: pattern', function validateMaxLength() {
+  it('should be valid when `value` matches `prop` regex', function () {
+    assert(pattern({}, '__matcher__', /.*matcher.*/) === false);
+  });
+  it('should be invalid when `value` does not match `prop` regex', function () {
+    assert(pattern({}, '__doesNotMatch__', /.*matcher.*/) === true);
+  });
+  it('should be invalid when `prop` is boolean', function () {
+    assert.throws(
+      () => pattern({}, 'z', true),
+      TypeError
+    );
+  });
+  it('should be invalid when `prop` is string', function () {
+    assert.throws(
+      () => pattern({}, 'z', 'a nice string'),
+      TypeError
+    );
+  });
+  it('should be invalid when `prop` is object', function () {
+    assert.throws(
+      () => pattern({}, 'z', {}),
+      TypeError
+    );
+  });
+  it('should be invalid when `prop` is number', function () {
+    assert.throws(
+      () => pattern({}, 'z', 99),
+      TypeError
+    );
+  });
+});

--- a/src/__tests__/promise.spec.js
+++ b/src/__tests__/promise.spec.js
@@ -1,0 +1,33 @@
+import assert from 'assert';
+import { promise } from '../basic-validations';
+
+// Note: each validator returns `true` when the `value` is invalid
+describe('Validator: promise', function () {
+  describe('Should throw Error', () => {
+    it('When `prop` is boolean', () => {
+      assert.throws(() => promise({}, '', true), Error);
+    });
+    it('When `prop` is string', () => {
+      assert.throws(() => promise({}, '', ''), Error);
+    });
+    it('When `prop` is object', () => {
+      assert.throws(() => promise({}, '', {}), Error);
+    });
+    it('When `prop` is number', () => {
+      assert.throws(() => promise({}, '', 46), Error);
+    });
+  });
+  describe('Should return expected result', () => {
+    it('When function is provided', (done) => {
+      const testField = 1;
+      const testValue = 2;
+      const testProp = (field, value) => {
+        assert(field === testField);
+        assert(value === testValue);
+        done();
+      };
+
+      promise(testField, testValue, testProp);
+    });
+  });
+});

--- a/src/__tests__/required.spec.js
+++ b/src/__tests__/required.spec.js
@@ -2,16 +2,16 @@ import assert from 'assert';
 import { required } from '../basic-validations';
 
 describe('Validator: required', function () {
-  it('should return false when `value` is empty and `prop` is false', function () {
+  it('should be valid when `value` is empty and `prop` is false', function () {
     assert(required({}, '', false) === false);
   });
-  it('should return false when `value` is truthy and `prop` is false', function () {
+  it('should be valid when `value` is truthy and `prop` is false', function () {
     assert(required({}, 'true', false) === false);
   });
-  it('should return true when `value` is empty and `prop` is true', function () {
+  it('should be invalid when `value` is empty and `prop` is true', function () {
     assert(required({}, '', true) === true);
   });
-  it('should return false when `value` is truthy and `prop` is true', function () {
+  it('should be valid when `value` is truthy and `prop` is true', function () {
     assert(required({}, 'true', true) === false);
   });
 });

--- a/src/__tests__/required.spec.js
+++ b/src/__tests__/required.spec.js
@@ -1,0 +1,17 @@
+import assert from 'assert';
+import { required } from '../basic-validations';
+
+describe('Validator: required', function validateRequired() {
+  it('should return false when `value` is empty and `prop` is false', function () {
+    assert(required({}, '', false) === false);
+  });
+  it('should return false when `value` is truthy and `prop` is false', function () {
+    assert(required({}, 'true', false) === false);
+  });
+  it('should return true when `value` is empty and `prop` is true', function () {
+    assert(required({}, '', true) === true);
+  });
+  it('should return false when `value` is truthy and `prop` is true', function () {
+    assert(required({}, 'true', true) === false);
+  });
+});

--- a/src/__tests__/required.spec.js
+++ b/src/__tests__/required.spec.js
@@ -1,7 +1,7 @@
 import assert from 'assert';
 import { required } from '../basic-validations';
 
-describe('Validator: required', function validateRequired() {
+describe('Validator: required', function () {
   it('should return false when `value` is empty and `prop` is false', function () {
     assert(required({}, '', false) === false);
   });

--- a/src/__tests__/setup.js
+++ b/src/__tests__/setup.js
@@ -1,0 +1,1 @@
+// test setup to be implemented

--- a/src/basic-validations.js
+++ b/src/basic-validations.js
@@ -56,7 +56,7 @@ export default {
     }
     // accept only spaces, digits and dashes
     if (/[^0-9 \-]+/.test(value)) {
-      return false;
+      return true;
     }
     var nCheck = 0,
       nDigit = 0,
@@ -66,7 +66,7 @@ export default {
     value = value.replace(/\D/g, "");
 
     // Basing min and max length on
-    // http://developer.ean.com/general_info/Valid_Credit_Card_Types
+    // http://developer.ean.com/general-info/valid-card-types/
     if (value.length < 13 || value.length > 19) {
       return false;
     }
@@ -83,7 +83,7 @@ export default {
       bEven = !bEven;
     }
 
-    return ( nCheck % 10 ) === 0;
+    return ( nCheck % 10 ) !== 0;
   }
 
 }

--- a/src/basic-validations.js
+++ b/src/basic-validations.js
@@ -48,7 +48,7 @@ export default {
     throw new Error("FormValidation: type promise must be a function!")
   },
   digits: function (field, value) {
-    return !field || /^\d+$/.test(value);
+    return !field || !/^\d+$/.test(value);
   },
   creditcard: function (field, value, prop) {
     if (!value) {

--- a/src/basic-validations.js
+++ b/src/basic-validations.js
@@ -28,7 +28,7 @@ export default {
   },
 
   equalTo: function (field, value, prop) {
-    return !value ? false : !prop == value;
+    return !value ? false : prop != value;
   },
 
   oneOf: function (field, value, prop) {

--- a/src/basic-validations.js
+++ b/src/basic-validations.js
@@ -4,7 +4,7 @@ export default {
   },
 
   minLength: function (field, value, prop) {
-    return prop && value ? value.length <= prop : false;
+    return prop && value ? value.length < prop : false;
   },
 
   maxLength: function (field, value, prop) {

--- a/src/basic-validations.js
+++ b/src/basic-validations.js
@@ -20,7 +20,7 @@ export default {
   },
 
   max: function (field, value, prop) {
-    return !isFinite(value) || parseFloat(value) <= prop;
+    return prop && value ? !isFinite(value) || parseFloat(value) > prop : false;
   },
 
   pattern: function (field, value, prop) {

--- a/src/basic-validations.js
+++ b/src/basic-validations.js
@@ -8,7 +8,7 @@ export default {
   },
 
   maxLength: function (field, value, prop) {
-    return prop && value ? value.length >= prop : false;
+    return prop && value ? value.length > prop : false;
   },
 
   email: function (field, value, prop) {

--- a/src/basic-validations.js
+++ b/src/basic-validations.js
@@ -1,3 +1,5 @@
+import validUrl from 'valid-url';
+
 export default {
   required: function (field, value, prop) {
     return prop ? !value : false
@@ -36,7 +38,7 @@ export default {
   },
 
   url: function (field, value, prop) {
-    return value ? false : !(/^(?:(?:(?:https?|ftp):)?\/\/)(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)(?:\.(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)*(?:\.(?:[a-z\u00a1-\uffff]{2,})).?)(?::\d{2,5})?(?:[/?#]\S*)?$/i.test(value))
+    return !value ? false : !validUrl.isUri(value);
   },
 
   promise: function (field, value, prop) {

--- a/src/basic-validations.js
+++ b/src/basic-validations.js
@@ -16,7 +16,7 @@ export default {
   },
 
   min: function (field, value, prop) {
-    return !isFinite(value) || parseFloat(value) >= prop;
+    return prop && value ? !isFinite(value) || parseFloat(value) < prop : false;
   },
 
   max: function (field, value, prop) {

